[AMDGPU] Improve isBasicBlockPrologue helper function

cdevadas opened the pull request:
This function helps to skip over the basic block prolog
instructions while inserting a new spill or a copy during
liverange split. However, this appears to be incomplete.
It currently skips only the instructions that restores
the exec mask. It should also have skipped over the BB
prolog spills and the liverange split copies inserted
earlier during regalloc while dealing with the BBLiveIns.

This patch partially addresses the limitation by including
only the spills in the query. The live-range-split COPY
handling is pending.

This helper function is also used at multiple places to
skip over the bb prolog. They should continue to produce
functionally correct code even after this patch.

This patch depends on [#69923](https://github.com/llvm/llvm-project/pull/69923)

cdevadas committed:
[AMDGPU] Try to fix the block prologs broken by RA inserted instructions

The insertion point determined by RA while attempting spills and liverange
split at the beginning of a block goes wrong at times, and the newly
inserted vector instructions are placed before the exec-mask restore
instruction which is wrong. It occurs mainly due to the dependency on
isBasicBlockPrologue that doesn't account early inserted instructions
(spills and splits) during RA and causes the block prolog break.

A better approach for deciding the insertion point should be worked out.
For now, improving the helper function to consider all possible early
insertions. This patch includes the spill instructions. The copies
associated with liverange split should also be included in the block
prolog.

ruiling commented:
Would you mind giving more details on the issue you are trying to solve? Did you hit the problem in any real case? Does this change has kind of relation with the change: https://reviews.llvm.org/D145329? Thanks!

cdevadas commented:
> Would you mind giving more details on the issue you are trying to solve? Did you hit the problem in any real case? Does this change has kind of relation with the change: https://reviews.llvm.org/D145329? Thanks!

Yes, this patch would avoid https://reviews.llvm.org/D145329.
Splitting the si_end_cf block doesn't effectively solve the problem in all cases. OCL conformance tests break with incorrect spill insertion in the downstream compiler. D145329, in fact, fixed it but created an incorrect spill placement for a math library code. Because the split inserts the *_term version of the exec manipulation instruction in the block which is a terminator and causes the spill insertion before it for the offending test. That is wrong.
The root cause I believe `isBasicBlockPrologue` isn't complete. Fixing this helper function avoids the need for end_cf block split.
Most of the lit test changes in the patch now insert the spills after the exec manipulation instruction, but that wasn't the case earlier.

arsenm commented:
I think isBasicBlockPrologue is unworkable and should be removed. I don't think it's sensible to do ad-hoc exec mask tracking like this and we need a better control flow representation. I think for the moment we are better off splitting blocks, and then starting to augment MachineBasicBlock with an annotation of the divergent CFG.

cdevadas commented:
> I think isBasicBlockPrologue is unworkable and should be removed. I don't think it's sensible to do ad-hoc exec mask tracking like this and we need a better control flow representation. I think for the moment we are better off splitting blocks, and then starting to augment MachineBasicBlock with an annotation of the divergent CFG.

The spills inserted at the block entry for BBLiveIns are part of the BB prolog. That's what I believe. So, teaching `isBasicBlockPrologue` seems the right approach to me unless there is a strong reason against it.
The initially proposed bb split approach still needs a fixup in the isBasicBlockPrologue. Any subsequent spill insertion in the same block causes a problem because the first spill would break the bb prolog which is wrong.

ruiling commented:
I agree that block prologue is hard to get correct behavior within LLVM. But as we strongly depend on it now, we cannot remove it before we came up with alternative solution. I think one possible way to remove block prologue is: don't reconverge the threads at block header, instead *always* setup the exec correctly in the terminators of predecessor block. But there are still many details to work out.

I don't have strong preference which way we should go (either always split block or the change here). But it is important it should work in all cases we have seen (e.g. livenrange-split might still insert vector instruction before end_cf even it is a terminator.)

We definitely want a test for such change, a '.ll' test is always helpful for revisiting the issue again.

cdevadas commented:
Rebased to reflect the pre-patches.

arsenm commented on llvm/lib/Target/AMDGPU/SIInstrInfo.cpp:
```diff
@@ -8335,8 +8335,14 @@ unsigned SIInstrInfo::getLiveRangeSplitOpcode(Register SrcReg,
-  return !MI.isTerminator() && MI.getOpcode() != AMDGPU::COPY &&
-         MI.modifiesRegister(AMDGPU::EXEC, &RI);
+  // It should cover all possible instructions inserted during regalloc in the
+  // block prolog.
```
This should elaborate on why. It's still a pretty dicey assumption

arsenm reviewed (changes requested):
I think the title needs to be reworked. "Improved" doesn't really explain what this does

ruiling commented:
> This patch passed in our internal testing (PSDB) and landing it would unblock my ongoing work related to wwm-allocation and a major issue observed in a customer code.

ok, sounds fine to me to solve the liverange-issue separately.

> Like I said earlier, we don't have a ready .ll test. I will try to get one forcing some reg-constraints.

But I think you should have a test that would fail without this change (not the liverange-split related issue) and pass with this change?

cdevadas commented:
> But I think you should have a test that would fail without this change (not the liverange-split related issue) and pass with this change?
I thought you were talking about the test case for liverange-split. I have a test for the SPILL issue. I can pre-commit the test first with the original issue indicated in it.

cdevadas commented:
Rebase + Addressed review comments.

arsenm commented on llvm/lib/Target/AMDGPU/SIInstrInfo.cpp:
```diff
@@ -8430,8 +8442,16 @@ unsigned SIInstrInfo::getLiveRangeSplitOpcode(Register SrcReg,
-  return !MI.isTerminator() && MI.getOpcode() != AMDGPU::COPY &&
-         MI.modifiesRegister(AMDGPU::EXEC, &RI);
+  // It should cover all possible instructions (spills, copies, etc.) early
+  // inserted during RA in the block prolog to avoid the broken prologs that
```
I think the "broken prolog" reads a bit too vaguely. Also not sure what "early" is doing here. How about

"We need to handle instructions which may be inserted during register allocation to handle the prolog. The initial prolog instruction may have been separated from the start of the block by spills and copies inserted needed by the prolog"

I still don't feel particularly comfortable trusting any spills here were for the prolog

cdevadas replied:
Done

arsenm reviewed (approved):
We really need something properly embedded in the IR for this but I guess we can go with this for now

piotrAMD commented:
Starting from this commit we are hitting the following assertion when running Vulkan CTS:

lib/CodeGen/SplitKit.cpp:1648: void llvm::SplitEditor::splitLiveThroughBlock(unsigned int, unsigned int, llvm::SlotIndex, unsigned int, llvm::SlotIndex): Assertion `(!LeaveBefore || Idx <= LeaveBefore) && "Interference"' failed.

I'll prepare a reproducer.

piotrAMD commented:
[repro.txt](https://github.com/llvm/llvm-project/files/13208325/repro.txt)

llc_bad -mtriple=amdgcn--amdpal -march=amdgcn -mcpu=gfx1030 repro.txt

piotrAMD commented:
Ping. Is this being looked into?
We have reverted the change in the graphics downstream branch, but the issue exists in trunk.

cdevadas commented:
> Ping. Is this being looked into? We have reverted the change in the graphics downstream branch, but the issue exists in trunk.

I'm looking into it. The assertion started coming because the insertion point is now modified. SplitKit expects the COPY during liverange split to be inserted at the top of the BB. But after my patch, the insertion point vary and they might not be always inserted before the first instruction. One or the other way, the greedy and Fastalloc have problem with the recomputed bb-prolog. I'm trying to see if the insertion point can be retained for SGPR spills/copies as an immediate fixup. But isBasicBlockPrologue being a common interface used for InlineSpiller, SplitKit and MachineIRSink to identify the insertion point. It is becoming difficult to handle the case.  I will provide an update soon.

cdevadas commented:
> [repro.txt](https://github.com/llvm/llvm-project/files/13208325/repro.txt)
> 
> llc -mtriple=amdgcn--amdpal -march=amdgcn -mcpu=gfx1030 repro.txt

@piotrAMD https://github.com/llvm/llvm-project/pull/72140 would fix it. I merged it to the upstream.
